src/admin_ws.rs

```diff
@@ -0,0 +1,187 @@
+/// when the connection is lost due to network issues or other failures.
+#[derive(Clone)]
+pub struct ReconnectingAdminWebsocket {
+    /// The WebSocket URL to connect to
+    url: String,
+    /// The handle to the AdminWebsocket connection
+    connection_handle: Arc<Mutex<Option<AdminWebsocket>>>,
+    /// Current retry attempt counter
+    current_retries: usize,
+}
+
+impl ReconnectingAdminWebsocket {
+    /// Creates a new ReconnectingAdminWebsocket with the specified parameters
+    pub fn new(url: &str) -> Self {
+        ReconnectingAdminWebsocket {
+            url: url.to_string(),
+            connection_handle: Arc::new(Mutex::new(None)),
+            current_retries: 0,
+        }
+    }
+
+    /// Checks if there is an active connection
+    fn is_connected(&self) -> HcHttpGatewayResult<bool> {
+        let connection = self.connection_handle.lock().map_err(|e| {
+            HcHttpGatewayError::InternalError(format!(
+                "Mutex was poisoned during is_connected check: {}",
+                e
+            ))
+        })?;
+
+        Ok(connection.is_some())
+    }
+
+    /// Gets reconnection retry count
+    pub fn get_reconnection_retries(&self) -> usize {
+        self.current_retries
+    }
+
+    /// Ensures that a connection is established before proceeding
+    async fn ensure_connected(&mut self) -> HcHttpGatewayResult<()> {
+        if self.is_connected()? {
+            return Ok(());
+        }
+
+        self.connect().await
+    }
+
+    /// Attempts to connect to the AdminWebsocket
+    ///
+    /// This will make multiple attempts according to the `max_retries` and
+    /// `retry_delay_ms` settings, with exponential backoff between retries.
+    pub async fn connect(&mut self) -> HcHttpGatewayResult<()> {
+        self.current_retries = 0;
+
+        while self.current_retries < ADMIN_WS_CONNECTION_MAX_RETRIES {
+            match AdminWebsocket::connect(&self.url).await {
+                Ok(conn) => {
+                    let mut connection = self.connection_handle.lock().map_err(|e| {
+                        HcHttpGatewayError::InternalError(format!(
+                            "Mutex was poisoned during is_connected check: {}",
+                            e
+                        ))
+                    })?;
+
+                    *connection = Some(conn);
+                    self.current_retries = 0;
+
+                    return Ok(());
+                }
+                Err(e) => {
+                    self.current_retries += 1;
+                    tracing::warn!(
+                        "Failed to connect to WebSocket (attempt {}/{}): {:?}",
+                        self.current_retries,
+                        ADMIN_WS_CONNECTION_MAX_RETRIES,
+                        e
+                    );
+
+                    if self.current_retries < ADMIN_WS_CONNECTION_MAX_RETRIES {
+                        sleep(Duration::from_millis(ADMIN_WS_CONNECTION_RETRY_DELAY_MS)).await;
```

Comment:
Why delay between connection attempts? There is an HTTP request waiting for this to happen. We should be making one attempt to connect if we ran into an error and otherwise we just return an error and free up resources to process a new request.